k8s/src/main/scala/io/buoyant/k8s/Watchable.scala

```diff
@@ -0,0 +1,122 @@
+
+/**
+ * An abstract class that encapsulates the ability to Watch a k8s [[Resource]].
+ */
+private[k8s] abstract class Watchable[O <: KubeObject: Manifest, W <: Watch[O]: Manifest]
+  extends Resource {
+  protected def backoffs: Stream[Duration]
+  protected def path: String
+  protected def stats: StatsReceiver
+
+  /**
+   * implementing classes should define this method to retrieve the current state
+   * of the resource to pass as Modified results to any watchers.
+   *
+   * @return a Future containing a sequence of Watches and an optional String representing the current resourceVersion
+   */
+  protected def restartWatches(
+    labelSelector: Option[String] = None,
+    fieldSelector: Option[String] = None
+  ): Future[(Seq[W], Option[String])]
+
+  protected def infiniteRetryFilter = new RetryFilter[http.Request, http.Response](
+    RetryPolicy.backoff(backoffs) {
+      // We will assume 5xx are retryable, everything else is not for now
+      case (_, Return(rep)) => rep.status.code >= 500 && rep.status.code < 600
+      case (_, Throw(NonFatal(ex))) =>
+        log.error(s"retrying k8s endpoints request on error $ex")
+        true
+    },
+    HighResTimer.Default,
+    stats,
+    RetryBudget.Infinite
+  )
+
+  /**
+   * Watch this resource for changes, using a chunked HTTP request.
+   *
+   * TODO: investigate k8s websockets support as an alternative HTTP approach
+   *
+   * @return An asynchronous stream of [[Watch]] objects containing additions, modifications, and
+   *         deletions to the items in the list being watched, and a Closable handle allowing for
+   *         the watch to be terminated.
+   */
+  def watch(
+    labelSelector: Option[String] = None,
+    fieldSelector: Option[String] = None,
+    resourceVersion: Option[String] = None
+  ): (AsyncStream[W], Closable) = {
+    val close = new AtomicReference[Closable](Closable.nop)
+
+    // Internal method used to recursively retry watches as needed on failures.
+    def _watch(resourceVersion: Option[String] = None): AsyncStream[W] = {
+      val req = Api.mkreq(http.Method.Get, path, None,
+        "watch" -> Some("true"),
+        "labelSelector" -> labelSelector,
+        "fieldSelector" -> fieldSelector,
+        "resourceVersion" -> resourceVersion)
+
+      def watchFromResponse(rsp: Response): AsyncStream[W] = rsp.status match {
+        // NOTE: 5xx-class statuses will be retried by the infiniteRetryFilter above.
+        case http.Status.Ok =>
+          close.set(Closable.make { _ =>
+            log.debug("k8s watch closed")
+            Future {
+              rsp.reader.discard()
+            } handle {
+              case _: Reader.ReaderDiscarded =>
+            }
+          })
+          val watchStream = Json.readStream[W](rsp.reader, Api.BufSize)
+          watchStream ++
+            AsyncStream.fromFuture(
+              watchStream.toSeq.map(_.lastOption.flatMap(_.resourceVersion)).flatMap {
```

Comment:
might factor out a helper method `def last[T](as: AsyncStream[T]): Option[T]` with a comment here about why it's safe to force the stream